TROPO ISO XML Generation

kevinmarlis opened the pull request:
## Description
- Adds support for ISO XML and Measured Parameters Config for TROPO PGE
- Ensures that framework for populating ISO XML with TROPO output product metadata works as intended.
- There are a few outstanding values that are using placeholders until we get confirmation otherwise:
    - latitude / longitude spacing: Product spec says ~0.07 degree resolution which is equivalent to roughly 8000 meters
    - spatial extent: Product is global on a regular 2d grid, so we're using the min/max lat/lon values to define the spatial extent
    - zone identifier: No reference in product spec, but the product is global
    - start_time / end_time: Single time value per product along with a temporal_resolution metadata field of "6h". Currently using time value from product for start_time and start_time + timedelta(hours=6) for end_time

## Affected Issues
- Closes #654 

## Testing
- Added ISO XML test which ensures all placeholders have been filled and that the appropriate error is raised when unable to generate ISO XML
- Will update after running Build/Test Jenkins pipeline

RKuttruff commented:
@kevinmarlis Secrets detection is flagging this PR, which is normal for ISO XML additions because they contain contact info. I'll message you with the details

RKuttruff commented on src/opera/pge/tropo/templates/OPERA_ISO_metadata_L4_TROPO_template.xml.jinja2:
```diff
@@ -0,0 +1,720 @@
+                    <gmd:authority>
+                        <gmd:CI_Citation>
+                            <gmd:title>
+                                <gco:CharacterString>Universal Transverse Mercator (UTM)</gco:CharacterString>
```
The product spec says this is in WGS84 reference system. I don't know if that requires changing this or how to properly cite for WGS84

kevinmarlis replied:
Ah, good catch! I believe it should be 
```<gco:CharacterString>World Geodetic System 1984 (WGS84)</gco:CharacterString>```

collinss-jpl commented on src/opera/util/h5_utils.py:
```diff
@@ -141,6 +141,92 @@ def convert_h5py_dataset(dataset_object):
     return result
 
 
+def get_hdf5_attrs_as_dict(file_name, group_path, ignore_keys=[]):
```
Usually want to avoid assigning a mutable object as a default argument value. `ignore_keys` should default to `None` and this function should initialize it to an empty list if `None` is provided.

kevinmarlis replied:
Great catch, thank you!

RKuttruff commented on src/opera/pge/tropo/templates/OPERA_ISO_metadata_L4_TROPO_template.xml.jinja2:
```diff
@@ -0,0 +1,720 @@
+                        <gco:Integer>{{ product_output.xCoordinates.size }}{# ISO_OPERA_rangeCount #}</gco:Integer>
+                    </gmd:dimensionSize>
+                    <gmd:resolution>
+                        <gco:Length uom="meter">{{ product_output.xCoordinates.spacing }}{# ISO_OPERA_rangePixelSize #}</gco:Length>
```
I couldn't find it in the schema on a quick search, but I wonder if this can be given in degrees instead of meters

kevinmarlis replied:
I think we can use degree but would need to change Length to Angle as well. I can try and track down some more resources on this.

RKuttruff replied:
FWIW I tried pulling the ISO files for global WGS84 collections like MUR, and they didn't even have some of these sections like spacing and zone identifier

collinss-jpl reviewed (changes requested):
Branch is looking good, just a few minor things to address

RKuttruff commented on src/opera/pge/tropo/templates/OPERA_ISO_metadata_L4_TROPO_template.xml.jinja2:
```diff
@@ -0,0 +1,720 @@
+                <gco:CharacterString>Data product generated in Network Common Data From version 4 format with ISO 19115 conformant metadata.</gco:CharacterString>
+            </gmd:environmentDescription>
+            <!-- This section documents the granules spatial and temporal extent as well as the grid mapping names and the projection names - NOT the TilingIdentificationSystem. -->
+            <gmd:extent>
```
Should we include vertical extent here since this product exists at multiple elevation layers?

kevinmarlis replied:
We can, but the extent is variable as the vertical layers produced are configurable via the runconfig.

RKuttruff commented on src/opera/pge/tropo/templates/OPERA_ISO_metadata_L4_TROPO_template.xml.jinja2:
```diff
@@ -0,0 +1,720 @@
+                    <gmd:distributionFormat>
+                        <gmd:MD_Format>
+                            <gmd:name>
+                                <gco:CharacterString>GeoTIFF</gco:CharacterString>
```
This should be NetCDF

RKuttruff commented on src/opera/pge/tropo/tropo_pge.py:
```diff
@@ -150,7 +152,254 @@ def _checksum_output_products(self):
+            'ISO_OPERA_ProducerGranuleId': granule_filename,
+            'MetadataProviderAction': "creation",
+            'GranuleFilename': granule_filename,
+            'ISO_OPERA_ProjectKeywords': ['OPERA', 'JPL', 'DSWx', 'Dynamic', 'Surface', 'Water', 'Extent'],
```
Should update these to apply to TROPO

RKuttruff commented on src/opera/test/pge/tropo/test_tropo_pge.py:
```diff
@@ -121,6 +126,16 @@ def test_tropo_pge_execution(self):
+            pge.runconfig.output_product_path, pge._catalog_metadata_filename())
+        self.assertTrue(os.path.exists(expected_catalog_metadata_file))
+
+        # # Check that the ISO metadata file was created and filled in as expected
+        expected_iso_metadata_file = join(
+            pge.runconfig.output_product_path, pge._iso_metadata_filename())
+        self.assertTrue(os.path.exists(expected_iso_metadata_file))
```
You should add a check here that the ISO file is fully populated and does not contain the template undefined error string
Reference: https://github.com/nasa/opera-sds-pge/blob/f13869f9c67a1b3451c3ca75fe10bcfcb09d96dc/src/opera/test/pge/dist_s1/test_dist_s1_pge.py#L179-L182

RKuttruff reviewed (changes requested):
I went over the ISO template and raised some questions, but also found some things to change/add

kevinmarlis commented on src/opera/pge/tropo/tropo_pge.py:
```diff
@@ -150,7 +152,254 @@ def _checksum_output_products(self):
+        nc_product = None
+        for output_product in output_products:
+            if output_product.endswith('nc'):
+                nc_product = output_product
```
Even though we're not currently doing any file renaming for TROPO, this for loop does include that step. I could break out the '.nc' search into it's own for loop and add the break.

kevinmarlis commented on src/opera/util/h5_utils.py:
```diff
@@ -141,6 +141,92 @@ def convert_h5py_dataset(dataset_object):
+        group_object = h5file.get(group_path)
+
+        if group_object is None:
+            raise RuntimeError(f"An error occurred retrieving object '{group_path}' "
```
I'll go ahead and add it to `get_hdf5_group_as_dict` as well since I'm in here already

kevinmarlis commented:
Addressed concerns and closed conversations where changes were implemented. Test is passing on opera-dev-pge. Outstanding items are:

- ISO XML WGS84 usage of spacing and zone identifier
- ISO XML addition of vertical extent

RKuttruff reviewed (commented):
All my main requested changes have been resolved. So I think this is good to go pending discussion on the outstanding points/questions Kevin mentioned.
